transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java

```diff
@@ -93,132 +108,287 @@ public Channel channel() {
+                @Override
+                public void run() {
+                    synchronized (DefaultChannelPipeline.this) {
+                        addLast0(newCtx);
+                    }
+                    callHandlerAdded0(newCtx);
+                }
+            }));
+        }
         return this;
     }
 
     private void addLast0(AbstractChannelHandlerContext newCtx) {
-        checkMultiplicity(newCtx);
-
         AbstractChannelHandlerContext prev = tail.prev;
         newCtx.prev = prev;
         newCtx.next = tail;
         prev.next = newCtx;
         tail.prev = newCtx;
-
-        callHandlerAdded(newCtx);
     }
 
     @Override
     public ChannelPipeline addBefore(String baseName, String name, ChannelHandler handler) {
-        return addBefore((ChannelHandlerInvoker) null, baseName, name, handler);
+        return addBefore(null, null, baseName, name, handler);
     }
 
     @Override
-    public synchronized ChannelPipeline addBefore(
+    public ChannelPipeline addBefore(
             EventExecutorGroup group, String baseName, String name, ChannelHandler handler) {
-        AbstractChannelHandlerContext ctx = getContextOrDie(baseName);
-        name = filterName(name, handler);
-        addBefore0(ctx, new DefaultChannelHandlerContext(this, findInvoker(group), name, handler));
-        return this;
+        return addBefore(group, null, baseName, name, handler);
     }
 
     @Override
-    public synchronized ChannelPipeline addBefore(
+    public ChannelPipeline addBefore(
             ChannelHandlerInvoker invoker, String baseName, String name, ChannelHandler handler) {
-        AbstractChannelHandlerContext ctx = getContextOrDie(baseName);
-        name = filterName(name, handler);
-        addBefore0(ctx, new DefaultChannelHandlerContext(this, invoker, name, handler));
-        return this;
+        return addBefore(null, invoker, baseName, name, handler);
     }
 
-    private void addBefore0(AbstractChannelHandlerContext ctx, AbstractChannelHandlerContext newCtx) {
-        checkMultiplicity(newCtx);
+    private ChannelPipeline addBefore(EventExecutorGroup group,
+            ChannelHandlerInvoker invoker, String baseName, String name, ChannelHandler handler) {
+        assertGroupAndInvoker(group, invoker);
+
+        final EventExecutor executor;
+        final AbstractChannelHandlerContext newCtx;
+        final AbstractChannelHandlerContext ctx;
+        final boolean inEventLoop;
+        synchronized (this) {
+            checkMultiplicity(handler);
+            ctx = getContextOrDie(baseName);
 
+            if (group != null) {
+                invoker = findInvoker(group);
+            }
+
+            newCtx = new DefaultChannelHandlerContext(this, invoker, filterName(name, handler), handler);
+            executor = executorSafe(invoker);
+
+            // If the executor is null it means that the channel was not registered on an eventloop yet.
```

Comment:
make these comments consistent throughout this PR with the javadoc comments below.